cupy/core/_routines_math.pyx

```diff
@@ -216,32 +372,126 @@ cdef ndarray scan(ndarray a, ndarray out=None):
+    padded_size = a.size // batch_size * padded_bs
+
+    kern_scan = _inclusive_batch_scan_kernel(a.dtype, block_size, op)
+    kern_scan(grid=((padded_size - 1) // (block_size) + 1,),
+              block=(block_size,),
+              args=(a, out, batch_size),
+              shared_mem=a.itemsize * block_size)
+    if batch_size > block_size:
+        blocked_sum = out[:, block_size-1::block_size]
+        _batch_scan_op(blocked_sum, op, dtype, blocked_sum)
+        kern_add = _add_scan_batch_blocked_sum_kernel(
+            out.dtype, op, block_size)
+        kern_add(
+            grid=((out.size - 1) // (block_size) + 1,),
+            block=(block_size,),
+            args=(out, batch_size))
+    return out
+
+
+cdef _axis_to_first(ndarray x, axis):
+    trans = [axis] + [a for a in range(x.ndim) if a != axis]
+    pre = list(range(1, axis + 1))
+    succ = list(range(axis + 1, x.ndim))
+    revert = pre + [0] + succ
+    return trans, revert
+
+
+cdef _proc_as_batch(ndarray x, axis, dtype, op):
+    if x.shape[axis] == 0:
+        return cupy.empty_like(x)
+    trans, revert = _axis_to_first(x, axis)
+    t = x.transpose(trans)
+    s = t.shape
+    r = t.reshape(x.shape[axis], -1)
+    # This is to use the current implemented fast scan
+    # TODO(ecastill) merge with above transformations
+    r = r.transpose((1, 0))
+    r = _batch_scan_op(r, op, dtype, r)
+    r = r.transpose((1, 0))
+    return r.reshape(s).transpose(revert)
+
+
+cpdef scan_core(ndarray a, axis, op, dtype=None, ndarray out=None):
+    if out is None:
+        if dtype is None:
+            kind = a.dtype.kind
+            if kind == 'b':
+                dtype = numpy.dtype('l')
+            elif kind == 'i' and a.dtype.itemsize < numpy.dtype('l').itemsize:
+                dtype = numpy.dtype('l')
+            elif kind == 'u' and a.dtype.itemsize < numpy.dtype('L').itemsize:
+                dtype = numpy.dtype('L')
+            else:
+                dtype = a.dtype
+        result = a.astype(dtype=dtype)
+    else:
+        if (out.flags.c_contiguous or out.flags.f_contiguous):
+            result = out
+        else:
+            result = a.astype(out.dtype)
+        result[...] = a
+    if axis is None:
+        result = result.ravel()
+        if cupy.cuda.cub_enabled:
+            # result will be None if the scan is not compatible with CUB
+            if op == scan_op.SCAN_SUM:
+                cub_op = cub.CUPY_CUB_CUMSUM
+            else:
+                cub_op = cub.cub.CUPY_CUB_CUMPROD
+            res = cub.cub_scan(result, cub_op)
+            if res is not None:
+                return res
+        scan(result, op, dtype, result)
+    else:
+        axis = cupy.util._normalize_axis_index(axis, a.ndim)
+        result = _proc_as_batch(result, axis, dtype, op)
+    if out is not None and out.data != result.data:
+        out[...] = result.reshape(out.shape)
+    else:
+        out = result
+    return out
```

Comment:
I am a bit confused. I thought you would do an in-place scan, so is this check-and-copy necessary?

Reply:
This is in the case that the output array is non-contiguous. for contiguous out arguments the scan is in place

Reply:
Ah, I see, thanks for explanation!